include/envoy/config/xds_grpc_context.h

```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+#include "envoy/config/subscription.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Config {
+
+/**
+ * Interface to an object that uses a Config::GrpcStream to carry out the xDS protocol.
+ */
+class XdsGrpcContext {
+public:
+  virtual ~XdsGrpcContext() = default;
+
+  /**
+   * For the GrpcStream to prompt the context to take appropriate action in response to the
```

Comment:
Are these callbacks? Should this be called `XdsGrpcCallbacks` to match Envoy convention here?

Reply:
I'm thinking we would have some object like:
```
class GrpcStreamCallbacks {
  virtual void onStreamEstablished() PURE;
  virtual void onEstablishedmentFailure() PURE;
}
```

This is passed to the `GrpcStream` independent of the context. Maybe concrete instances of this are in fact the context etc., but I don't think the context needs to actually have these as part of its API.

FWIW, minior nit, but the `onFooBar()` naming is much more common in Envoy than `handleFooBar`, so best to stick with that.

Reply:
Ahh, yes, good idea - no reason not to have the context also have that callbacks-grouping interface. The need for GrpcStream to be able to call drainRequests as part of its rate limit logic made me think in terms of it calling into an actual object, but actually, that can be sensibly abstracted as a onWriteable(). Additionally, I have pulled the std::function for handling discovery responses out into the callbacks interface as well.

(Yeah, sorry about the naming; that was the result of evolving from functions that were called normally by other functions in the same class).